libusb/os/linux_usbfs.c

```diff
@@ -449,21 +486,160 @@ static void op_exit(struct libusb_context *ctx)
+	int r, has_usbhost;
+
+	struct android_jni_devices *devices;
+	jobject device;
+	uint8_t busnum, devaddr;
+
+	r = android_jni_detect_usbhost(cpriv->android_jni, &has_usbhost);
+
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	if (!has_usbhost) {
+		usbi_dbg("This device does not have the android.hardware.usb.host feature");
+		return LIBUSB_ERROR_NOT_SUPPORTED;
+	}
+
+	usbi_dbg("enumerating usb devices using android api");
+
+	r = android_jni_devices_alloc(cpriv->android_jni, &devices);
+
+	if (r != LIBUSB_SUCCESS)
+		return r;
+	
+	while (LIBUSB_SUCCESS ==
+		android_jni_devices_next(devices, &device, &busnum, &devaddr))
+	{
+
+		if (linux_enumerate_device(ctx, busnum, devaddr, NULL) < 0)
+			usbi_dbg("failed to enumerate android device %d/%d", busnum, devaddr);
+	
+		android_jni_globalunref(cpriv->android_jni, device);
+	}
+
+	android_jni_devices_free(devices);
+	
+	return LIBUSB_SUCCESS;
+}
+
+static int get_android_jni_fd(struct libusb_device_handle *handle)
+{
+	struct linux_context_priv *cpriv = usbi_get_context_priv(HANDLE_CTX(handle));
+	struct libusb_device *dev = handle->dev;
+	struct linux_device_priv *priv = usbi_get_device_priv(dev);
+	struct linux_device_handle_priv *hpriv = usbi_get_device_handle_priv(handle);
+
+	int r, fd;
+	uint8_t *descriptors;
+	size_t descriptors_len;
+
+	r = android_jni_connect(cpriv->android_jni, priv->android_jni_device,
+		&hpriv->android_jni_connection, &fd, &descriptors, &descriptors_len);
+	if (r != LIBUSB_SUCCESS) {
+		if (r == LIBUSB_ERROR_ACCESS)
+			android_jni_request_permission(cpriv->android_jni, priv->android_jni_device);
+		return r;
+	}
+
+	free(priv->descriptors);
+	priv->descriptors = descriptors;
+	priv->descriptors_len = descriptors_len;
+
+	/* right now android_jni_device is only filled if sysfs is not being used,
+	 * so localize the device descriptor as if this were usbfs */
+	usbi_localize_device_descriptor(priv->descriptors);
+
+	memcpy(&dev->device_descriptor, priv->descriptors, LIBUSB_DT_DEVICE_SIZE);
+	parse_config_descriptors(dev);
+
+	return fd;
+}
+
+static int android_jni_initialize_device(struct libusb_device *dev, uint8_t busnum,
+	uint8_t devaddr)
+{
+	struct libusb_context *ctx = DEVICE_CTX(dev);
+	struct linux_context_priv *cpriv = usbi_get_context_priv(ctx);
+	struct linux_device_priv *priv = usbi_get_device_priv(dev);
+
+	char *strings;
+	size_t strings_len;
+
+	struct android_jni_devices *devices;
+	jobject iter_device;
+	uint8_t iter_busnum, iter_devaddr;
+
+	int r = android_jni_devices_alloc(cpriv->android_jni, &devices);
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	priv->android_jni_device = NULL;
+
+	while (LIBUSB_SUCCESS ==
+		android_jni_devices_next(devices, &iter_device,
+		                         &iter_busnum, &iter_devaddr))
+	{
+		if (iter_busnum == busnum && iter_devaddr == devaddr) {
+			priv->android_jni_device = iter_device;
+			break;
+		} else {
+			android_jni_globalunref(cpriv->android_jni, iter_device);
+		}
+	}
+
```

Comment:
I believe `devices` isn't being freed in this function
```suggestion

	android_jni_devices_free(devices);

```